tc/core/tc_executor_new_api.h

```diff
@@ -0,0 +1,125 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include "tc/core/tc2halide.h"
+#include "tc/core/tensor.h"
+#include "tc/core/utils/time.h"
+#include "tc/lang/tree.h"
+
+namespace tc {
+/**
+ * TcExecutor is a backend-agnostic abstraction that provides the base
+ * functionality for an object returned by compilation to run.
+ * TcExecutor is templated by the Backend type.
+ * When a derived executor inherits from TcExecutor<Backend>, it
+ * provides support for running a compiled TC on the particular backend.
+ *
+ * A TcExecutor mixes templating and inheritance:
+ *   1. templating is necessary because we want type-safety with proper
+ *      dependent types (e.g. Backend::CompilationResultType)
+ *   2. inheritance is necessary because executors require specific
```

Comment:
I don't understand this motivation.  Inheritance is necessary when all derived
types need to have some _common_ information.  This comment makes it feel like
they need different information, so there would be no point for inheritance.